File: src/main/java/dev/budget/reconciler/model/MintTransaction.java
package dev.budget.reconciler.model;

import org.joda.time.LocalDate;

public class MintTransaction implements Transaction {

	private LocalDate date;
	private String description;
	private String originalDescription;
	private long amountCents;
	private String type;
	private String category;
	private String account;

	public MintTransaction() {}

	public MintTransaction(LocalDate date, String description, String originalDescription, int amountCents, String type, String category, String account) {
		this.date = date;
		this.description = description;
		this.originalDescription = originalDescription;
		this.amountCents = amountCents;
		this.type = type;
		this.category = category;
		this.account = account;
	}

	public LocalDate getDate() {
		return date;
	}

	public void setDate(LocalDate date) {
		this.date = date;
	}

	public String getDescription() {
		return description;
	}

	public void setDescription(String description) {
		this.description = description;
	}

	public String getOriginalDescription() {
		return originalDescription;
	}

	public void setOriginalDescription(String originalDescription) {
		this.originalDescription = originalDescription;
	}

	public long getAmountCents() {
		return amountCents;
	}

	public void setAmountCents(long amountCents) {
		this.amountCents = amountCents;
	}

	public String getType() {
		return type;
	}

	public void setType(String type) {
		this.type = type;
	}

	public String getCategory() {
		return category;
	}

	public void setCategory(String category) {
		this.category = category;
	}

	public String getAccount() {
		return account;
	}

	public void setAccount(String account) {
		this.account = account;
	}

	public String toString() {
		return "MintTransaction{" +
				"date=" + date +
				", description='" + description + '\'' +
				", originalDescription='" + originalDescription + '\'' +
				", amountCents=" + amountCents +
				", type='" + type + '\'' +
				", category='" + category + '\'' +
				", account='" + account + '\'' +
				'}';
	}
}

File: src/main/scala/dev/budget/reconciler/model/MintTransaction.scala
package dev.budget.reconciler.model

import java.io.IOException

import org.elasticsearch.common.xcontent.{XContentBuilder, XContentFactory}
import org.joda.time.LocalDate
import org.joda.time.format.DateTimeFormat
import org.slf4j.Logger

case class MintTransaction(
  date: LocalDate,
  description: String,
  originalDescription: String,
  amountCents: Long,
  transactionType: String,
  category: String,
  account: String) extends Transaction {

  private val log: Logger = org.slf4j.LoggerFactory.getLogger(getClass)

  override def esJson: Option[XContentBuilder] = {
    val dateStr: String = DateTimeFormat.forPattern("yyyy-MM-dd").print(date)

    try {
      Some(XContentFactory.jsonBuilder()
        .startObject
          .field("date", dateStr)
          .field("description", description)
          .field("originalDescription", originalDescription)
          .field("amountCents", amountCents)
          .field("transactionType", transactionType)
          .field("category", category)
          .field("account", account)
        .endObject
      )
    } catch {
      case e: IOException =>
        log.error(s"Error converting transaction $this to json")
        None
    }
  }
}
